operator/v1/types_ingress.go

```diff
@@ -1346,6 +1346,78 @@ type IngressControllerHTTPHeaders struct {
 	// +nullable
 	// +optional
 	HeaderNameCaseAdjustments []IngressControllerHTTPHeaderNameCaseAdjustment `json:"headerNameCaseAdjustments,omitempty"`
+
+	// actions specifies options for performing action on headers.
+	// For Setting Headers you can provide a value for the header whereas for deleting one needs to provide header name.
+	// +optional
+	Actions IngressControllerHTTPHeaderActions `json:"actions,omitempty"`
+}
+
+// IngressControllerHTTPHeaderActions defines whether we want to set/delete Response or Request Headers.
+type IngressControllerHTTPHeaderActions struct {
+	// Response is a list of http response headers which either need to set or deleted.
+	Response []IngressControllerHTTPHeader `json:"response,omitempty"`
+	// Request is a list of http request headers which either need to set or deleted.
+	Request []IngressControllerHTTPHeader `json:"request,omitempty"`
+}
+
+// IngressControllerSetOrDeleteHTTPHeader specifies options for set/delete a header.
+type IngressControllerHTTPHeader struct {
+	// name specifies a header name to be set/delete.  Its value must be a valid HTTP header.
+	// name as defined in RFC 2616 section 4.2.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=256
+	// +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+	Name string `json:"name"`
+
+	// value specifies fields to mention the header details and the action you want to perform on them.
+	Value IngressControllerHTTPHeaderActionUnion `json:"value"`
```

Comment:
@JoelSpeed has asked not to use inline.

Reply:
His explanation -
https://github.com/openshift/api/pull/1364/files#r1067006894
[8:02](https://redhat-internal.slack.com/archives/D03T78CTU8Y/p1673517759881229)
No, that's the whole point, if you use inline, you make the yaml look like one thing, which has caused confusion, so we want to make it explicit
[8:03](https://redhat-internal.slack.com/archives/D03T78CTU8Y/p1673517829259369)
With inline
headers:
 response:
 - name: X-Forwarded-For
   action: Set 
   set:
     value: foo
Without:
headers:
 response:
 - name: X-Forwarded-For
   value:
     action: Set 
     set:
       value: foo
[8:04](https://redhat-internal.slack.com/archives/D03T78CTU8Y/p1673517841989589)
So it's clear that the union is represented by value in this case

Reply:
> Did you mean to name this field "value"? This naming results in the following structure:
> 
> ```yaml
> spec:
>   httpHeaders:
>     actions:
>       request:
>       - name: foo
>         value:
>           action: Set
>           set:
>             value: bar
>       - name: baz
>         value:
>           action: Delete
> ```
> 
> This example looks weird because (a) you have "value" twice in the "Set" case, and (b) it suggests that only the value of the header will be deleted in the "Delete" case.
> 
> Using an inline union for `IngressControllerHTTPHeaderActionUnion` would make the API much nicer:
> 
> ```yaml
> spec:
>   httpHeaders:
>     actions:
>       request:
>       - name: foo
>         action: Set
>         set:
>           value: bar
>       - name: baz
>         action: Delete
> ```
> 
> I think someone had an objection to using inline, but I can't find the comment on this PR or the EP PR. If we cannot use an inline union, could we rename the fields to something less misleading and silly looking? For example:
> 
> ```yaml
> spec:
>   httpHeaders:
>     actions:
>       request:
>       - name: foo
>         action:
>           operator: Set
>           set:
>             value: bar
>       - name: baz
>         action:
>           operator: Delete
> ```
> 
> I don't much like "operator", but I like using "value" for the `IngressControllerHTTPHeaderActionUnion` field or using "action" in three places even less.

Yes lets not use operator. I am not getting any value or use from it. action: Set is more better as it is a union discriminator.

Reply:
using inline I agree that the CR will remove redundant key "value"

Reply:
I don't have a strong opinion on the naming, but the inline union comments are correct. We don't allow them in OpenShift because they can produce unclear APIs, it's a better distinction when they are explicit unions, so we will have to come up with some naming options that everyone can agree on

What about using `type: Set` and `type: Delete`? ie the type of the action
```
spec:
  httpHeaders:
    actions:
      request:
      - name: foo
        action:
          type: Set
          set:
            value: bar
      - name: baz
        action:
          type: Delete
```

Reply:
I like using "type" for the union discriminator.  It's less ambiguous, less repetitive, and more consistent with most other union types.

Reply:
But why do we need this extra line. I am not sure if this is looking fine.